### PR TITLE
added resubscribe ability

### PR DESCRIPTION
**Pull request by gstarikov**
bad, dirty hotfix
but now subscribers won't lose subscribe to queue in case of rabbit restart

### CODE REVIEW

**Comment by asim** [comment]
It's really difficult for me to accept a PR that has 4 commits with the comment "bad hot fix", one that says "done" and then a PR comment that says "bad, dirty hotfix". It's also a fairly bad hack. I understand the need to enable resubscription on the event of a connection loss but this requires significant work.

File: broker/rabbitmq/rabbitmq.go [hunk 1]
```diff
@@ -8,15 +8,21 @@ import (
 	"github.com/micro/go-micro/broker"
 	"github.com/micro/go-micro/cmd"
 	"github.com/streadway/amqp"
+	"sync"
```
Comment (asim) [inline broker/rabbitmq/rabbitmq.go]: Move stdlib packages to be with the other stdlib imports.

File: broker/rabbitmq/rabbitmq.go [hunk 2]
```diff
@@ -8,15 +8,21 @@ import (
...
+	"time"
 )
 
 type rbroker struct {
 	conn  *rabbitMQConn
 	addrs []string
 	opts  broker.Options
+	lock sync.Mutex
```
Comment (asim) [inline broker/rabbitmq/rabbitmq.go]: Call this `mtx`

File: broker/rabbitmq/rabbitmq.go [hunk 3]
```diff
@@ -8,15 +8,21 @@ import (
...
+	wg sync.WaitGroup
 }
 
 type subscriber struct {
+	lock sync.Mutex
```
Comment (asim) [inline broker/rabbitmq/rabbitmq.go]: Call this `mtx`

File: broker/rabbitmq/rabbitmq.go [hunk 4]
```diff
@@ -121,13 +122,57 @@ func (r *rbroker) Subscribe(topic string, handler broker.Handler, opts ...broker
...
-	go func() {
-		for d := range sub {
-			go fn(d)
-		}
-	}()
-
-	return &subscriber{ch: ch, topic: topic, opts: opt}, nil
+	sret := &subscriber{ch: nil, topic: topic, opts: opt, mayRun:true}
+
+	go func(s*subscriber) {
```
Comment (asim) [inline broker/rabbitmq/rabbitmq.go]: All this code inlined like this is very unclean. Its probably better to move it to a function that can then be executed here.

File: broker/rabbitmq/rabbitmq.go [hunk 5]
```diff
@@ -121,13 +122,57 @@ func (r *rbroker) Subscribe(topic string, handler broker.Handler, opts ...broker
...
+		for {
+			s.lock.Lock()
+			mayRun := s.mayRun
+			s.lock.Unlock()
+			if !mayRun {
+				break
+			}
+			r.lock.Lock()// когда-нибудь у нас будет больше 1 потока
```
Comment (asim) [inline broker/rabbitmq/rabbitmq.go]: Please us english comments on a separate line

File: broker/rabbitmq/rabbitmq.go [hunk 6]
```diff
@@ -121,13 +122,57 @@ func (r *rbroker) Subscribe(topic string, handler broker.Handler, opts ...broker
...
+			if !r.conn.connected {// может упасть с паникой если делать Consume без соединения
+				time.Sleep(1 * time.Second)
```
Comment (asim) [inline broker/rabbitmq/rabbitmq.go]: time.Sleep should actually be some sort of exponential backoff

File: broker/rabbitmq/rabbitmq.go (hunk 1 above)
Comment (gstarikov) [reply broker/rabbitmq/rabbitmq.go]: ok

File: broker/rabbitmq/rabbitmq.go (hunk 2 above)
Comment (gstarikov) [reply broker/rabbitmq/rabbitmq.go]: ok

File: broker/rabbitmq/rabbitmq.go (hunk 3 above)
Comment (gstarikov) [reply broker/rabbitmq/rabbitmq.go]: ok

File: broker/rabbitmq/rabbitmq.go (hunk 5 above)
Comment (gstarikov) [reply broker/rabbitmq/rabbitmq.go]: ok

File: broker/rabbitmq/rabbitmq.go (hunk 6 above)
Comment (gstarikov) [reply broker/rabbitmq/rabbitmq.go]: ok. changed to wait on channel + exp sleep

**Comment by gstarikov** [comment]
last commit fixed with all reviewed comments and by now it's in our production for 5 days already
may be its time to make merge?

**Comment by asim** [comment]
This PR is in a bad state, I need you to go away and redo this, with one commit message that makes sense, a dco sign off and without all the log lines. It looks like you're still in the process of debugging. I really don't like the open PR like this so closing until you come back with something solid.

**Pull request closed without merging**